File: Retos/Reto #29 - EL CARÁCTER INFILTRADO [Fácil]/ruby/test0n3.rb
# frozen_string_literal: true

#
# Crea una función que reciba dos cadenas de texto casi iguales,
# a excepción de uno o varios caracteres.
# La función debe encontrarlos y retornarlos en formato lista/array.
# - Ambas cadenas de texto deben ser iguales en longitud.
# - Las cadenas de texto son iguales elemento a elemento.
# - No se pueden utilizar operaciones propias del lenguaje
#   que lo resuelvan directamente.
#
# Ejemplos:
# - Me llamo mouredev / Me llemo mouredov -> ["e", "o"]
# - Me llamo.Brais Moure / Me llamo brais moure -> [" ", "b", "m"]
#

# class Sentence get different characters between two sentences
class Sentence
  attr_accessor :sentence1, :sentence2

  def initialize(sentence1, sentence2)
    @sentence1 = sentence1
    @sentence2 = sentence2
    @diff = []
  end

  def find_diff
    return false if @sentence1.length != @sentence2.length

    @sentence1.length.times do |i|
      @diff << @sentence2[i] if sentence1[i] != sentence2[i]
    end
    @diff
  end
end

p Sentence.new('Me llamo mouredev', 'Me llemo mouredov').find_diff
p Sentence.new('Me llamo.Brais Moure', 'Me llamo brais moure').find_diff
p Sentence.new('first sentence', 'second sentence').find_diff
p Sentence.new('first  sentence', 'second sentence').find_diff

File: Retos/Reto #29 - EL CARÁCTER INFILTRADO [Fácil]/javascript/test0n3.js
const findDiff = (sentence1, sentence2) => {
  let diff = [];
  if (sentence1.length != sentence2.length) {
    return false;
  }
  for (let i = 0; i < sentence1.length; i++) {
    if (sentence1[i] != sentence2[i]) {
      diff.push(sentence2[i]);
    }
  }
  return diff;
};

const tests = {
  input: [
    ["Me llamo mouredev", "Me llemo mouredov"],
    ["Me llamo.Brais Moure", "Me llamo brais moure"],
    ["first sentence", "second sentence"],
    ["first  sentence", "second sentence"],
  ],
  expecteds: [
    ["e", "o"],
    [" ", "b", "m"],
    false,
    ["s", "e", "c", "o", "n", "d"],
  ],
};

let errors = 0;
tests.input.forEach((test, index) => {
  const result = findDiff(test[0], test[1]);
  if (JSON.stringify(result) == JSON.stringify(tests.expecteds[index])) {
    console.log(
      `Success - input: ${test[0]}, ${test[1]}, expected: ${tests.expecteds[index]}`
    );
  }
  if (JSON.stringify(result) != JSON.stringify(tests.expecteds[index])) {
    errors += 1;
    console.log(
      `Error - input: ${test[0]},${test[1]}, expected: ${tests.expecteds[index]}, result: ${result}`
    );
  }
});
console.log(`Errors: ${errors}`);
